### PR TITLE
Basic Banning Implementation

### PR DESCRIPTION
**Pull request by Kantanomo**
Added very basic server banning based off a file in the root directory called "Player Bans.txt" the file can be modified and loaded if the server is closed. Otherwise two commands have been added to manage it "ban" and "unban". The command "ban" can be used if the client is connected or not connected to the server. If connected the server owner can use the command to ban them by username, otherwise they can use the IP. the unban command can only be used with the IP address.

NitroxServer.Communication.Packets.Processors.MultiplayerSessionReservationRequestProcessor
To append the IP to the reservation request.

 NitroxServer.Communication.NetworkingLayer.LiteNetLib.LiteNetLibServer 
 Adds a condtion that checks if the IP Is in the ban list

Added IPBanning.cs
A static class used for managing bans
Runs at NitroxServer.Server construction
-- Banning.IpBanning.Load();

Adds two console commands
Ban (player/IP)
unban (player/ip)

These commands modify a file created at runtime in the root directory
Player bans.txt

### CODE REVIEW

File: NitroxServer/Banning/IpBanning.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace NitroxServer.Banning
+{
+    public static class IpBanning
+    {
+        public static List<string> BannedIPs = new List<string>();
+        public static string BanFilePath { get; } = Path.GetFullPath(Path.Combine(Environment.GetEnvironmentVariable("NITROX_LAUNCHER_PATH") ?? "", "Player Bans.txt"));
```
Comment (Jannify) [inline NitroxServer/Banning/IpBanning.cs:13]: Maybe get the launcher path from [NitroxModel/Helper/NitroxAppData.cs.](https://github.com/SubnauticaNitrox/Nitrox/blob/master/NitroxModel/Helper/NitroxAppData.cs)

File: NitroxServer/Banning/IpBanning.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace NitroxServer.Banning
+{
+    public static class IpBanning
```
Comment (Jannify) [inline NitroxServer/Banning/IpBanning.cs:10]: Maybe make a interface (with a shared method "IsBanned") for more banning options in the future and instantiate it with AutoFac (no static class)

File: NitroxServer/Banning/IpBanning.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        public static List<string> BannedIPs = new List<string>();
+        public static string BanFilePath { get; } = Path.GetFullPath(Path.Combine(Environment.GetEnvironmentVariable("NITROX_LAUNCHER_PATH") ?? "", "Player Bans.txt"));
+
+        public static void Load()
+        {
+            if (File.Exists(BanFilePath))
+            {
+                var fileStrings = File.ReadAllLines(BanFilePath);
+                BannedIPs.AddRange(fileStrings);
+
```
Comment (Jannify) [inline NitroxServer/Banning/IpBanning.cs:21]: Minor: Blank line

File: NitroxServer/Communication/NetworkingLayer/LiteNetLib/LiteNetLibServer.cs [hunk 4]
```diff
@@ -79,7 +79,7 @@ private void OnPacketReceived(WrapperPacket wrapperPacket, NetPeer peer)
 
         public void OnConnectionRequest(ConnectionRequest request)
         {
-            if (server.PeersCount < maxConn)
+            if (server.PeersCount < maxConn && !Banning.IpBanning.BannedIPs.Contains(request.RemoteEndPoint.Address.ToString()))
```
Comment (Jannify) [inline NitroxServer/Communication/NetworkingLayer/LiteNetLib/LiteNetLibServer.cs:82]: Better to move the check to the class

File: NitroxServer/Communication/Packets/Processors/MultiplayerSessionReservationRequestProcessor.cs [hunk 5]
```diff
@@ -29,7 +29,7 @@ public override void Process(MultiplayerSessionReservationRequest packet, Nitrox
                 authenticationContext,
                 correlationId);
 
-            Log.Info($"Reservation processed successfully: Username: {packet.AuthenticationContext.Username} - {reservation}");
+            Log.Info($"Reservation processed successfully: Username: {packet.AuthenticationContext.Username} - {reservation} - IP Address: {connection.Endpoint.Address.ToString()}");
```
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/MultiplayerSessionReservationRequestProcessor.cs:32]: We don't want to write the IP address to the log files for security reasons, so if you think logging the address is needed here use "InfoSensitive" and refer to examples in our code base for implementation.

File: NitroxServer/ConsoleCommands/BanIpCommand.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+using NitroxServer.Serialization;
+
+namespace NitroxServer.ConsoleCommands
+{
+    internal class BanIpCommand : Command
+    {
+        private readonly EntitySimulation entitySimulation;
+        private readonly PlayerManager playerManager;
+
+        public BanIpCommand(PlayerManager playerManager, EntitySimulation entitySimulation) : base("ban", Perms.MODERATOR, "ban a player from the server", true)
```
Comment (Jannify) [inline NitroxServer/ConsoleCommands/BanIpCommand.cs:21]: Minor:
```suggestion
        public BanIpCommand(PlayerManager playerManager, EntitySimulation entitySimulation) : base("ban", Perms.MODERATOR, "Bans a player from the server", true)
```

File: NitroxServer/ConsoleCommands/BanIpCommand.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                Banning.IpBanning.AddNewBan(playerToBan.connection.Endpoint.Address.ToString());
+                playerManager.SendPacketToOtherPlayers(new Disconnect(playerToBan.Id), playerToBan);
+                SendMessage(args.Sender, $"The player {playerToBan.Name} has been banned");
+            }
+            else
+            {
+                foreach (Player playerToBan in playerManager.GetAllPlayers())
+                {
+                    if (playerToBan.connection.Endpoint.Address.ToString() == temp.ToString())
+                    {
```
Comment (Jannify) [inline NitroxServer/ConsoleCommands/BanIpCommand.cs:54]: To reduce duplicated code (L38-47 and L58-L66) I would prefer restricting the if to finding the player-object/address.

File: NitroxServer/ConsoleCommands/BanIpCommand.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+                Player playerToBan = args.Get<Player>(0);
+                playerToBan.SendPacket(new PlayerKicked(args.GetTillEnd(1)));
+                playerManager.PlayerDisconnected(playerToBan.connection);
+
+                List<SimulatedEntity> revokedEntities = entitySimulation.CalculateSimulationChangesFromPlayerDisconnect(playerToBan);
+                if (revokedEntities.Count > 0)
+                {
+                    SimulationOwnershipChange ownershipChange = new SimulationOwnershipChange(revokedEntities);
+                    playerManager.SendPacketToAllPlayers(ownershipChange);
+                }
```
Comment (Jannify) [inline NitroxServer/ConsoleCommands/BanIpCommand.cs:43]: Maybe just call `KickCommand.Execute()`? You can get the object with AutoFac.

File: NitroxServer/ConsoleCommands/UnBanPlayer.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+using NitroxServer.Serialization;
+
+namespace NitroxServer.ConsoleCommands
+{
+    internal class UnBanIpCommand : Command
+    {
+        private readonly EntitySimulation entitySimulation;
+        private readonly PlayerManager playerManager;
+
+        public UnBanIpCommand(PlayerManager playerManager, EntitySimulation entitySimulation) : base("unban", Perms.MODERATOR, "unban a player from the server", true)
```
Comment (Jannify) [inline NitroxServer/ConsoleCommands/UnBanPlayer.cs:21]: Minor:
```suggestion
        public UnBanIpCommand(PlayerManager playerManager, EntitySimulation entitySimulation) : base("unban", Perms.MODERATOR, "Unbans a player from the server", true)
```

File: NitroxServer/ConsoleCommands/UnBanPlayer.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+            AddParameter(new TypePlayer("name/ip", true));
+        }
+
+        protected override void Execute(CallArgs args)
+        {
+            System.Net.IPAddress temp;
+            if (!System.Net.IPAddress.TryParse(args.Get(0), out temp))
+            {
+                Player player = args.Get<Player>(0);
```
Comment (Jannify) [inline NitroxServer/ConsoleCommands/UnBanPlayer.cs:34]: There are many places where you check if the IP is valid or not. Maybe move that to `AddNewBan()`/`RemoveBan()` of `IpBanning.cs`.

**Review by Jannify — CHANGES_REQUESTED** [review]
Sry that the review took so long, after a few days it just slipped my thoughts :D
I have marked a few places where I think there are better solutions. In most cases it's probably because you don't know the code base well yet ;), but that comes with time and is probably also due to the poor and outdated docs.

**Comment by dartasen** [comment]
Looks like this code will not be maintained by the author, also the code is not in a good state to be merged in our code base.
I'm closing it for now.

**Pull request closed without merging**